feat(compiler): allow self-closing tags on custom elements

crisbeto opened the pull request:
Allows for self-closing tags to be used for non-native tag names, e.g. `<foo [input]="bar"></foo>` can now be written as `<foo [input]="bar"/>`. Native tag names still have to have closing tags.

Fixes #39525.

crisbeto commented on packages/compiler/src/ml_parser/parser.ts:
```diff
@@ -275,12 +275,6 @@ class _TreeBuilder {
     if (this._peek.type === TokenType.TAG_OPEN_END_VOID) {
       this._advance();
       selfClosing = true;
-      const tagDef = this.getTagDefinition(fullName);
```
I was a bit on the fence here whether to allow self-closing tags for everything or just for non-native tags. I ended up going for the former for consistency, but I'm open to suggestions.

alxhub replied:
I do think we should limit self-closing to non-native tags, and still enforce proper HTML spec for HTML tags.

e-oz commented:
I’m sorry if the answer to my question is in the source code: will the compiler expand self-closed tags to regular tags to don't break the specification for custom elements?

crisbeto commented:
Yes, `<foo/>` will be treated the same as `<foo></foo>`.

e-oz commented:
So they will not be added to the DOM as self-enclosed, amazing.

Yogu commented:
In the issue you linked, @IgorMinar explained why this feature could not "just be added" and outlined that there would be an impact to the ecosystem.

> There are also additional considerations around template editing, syntax highlighting, autocompletion, linting etc - all of these tools nowadays work with Angular because they already support standard html. Once we move away from the spec, there is a possibility that common developer workflows could be affected or outright broken.

As far as I understand, supporting self-closing tags (that [do not exist in HTML](https://developer.mozilla.org/en-US/docs/Glossary/Void_element#self-closing_tags)) would mean that it would no longer be possible to use HTML parsers like the DOM to parse Angular templates.

Did you evaluate the impact of this change and think about how it could be rolled out smoothly?

e-oz commented:
@Yogu from the other side, developers are not forced to use self-closed tags only. While tools are being updated to this change, we can still use our existing code and write regular tags. If Angular will not make a step in this direction, tools will never follow.

crisbeto commented:
Angular templates aren't being parsed by the DOM. We have our own parser and compiler that turn them into JavaScript function calls.

TillFProtzek commented:
> Angular templates aren't being parsed by the DOM. We have our own parser and compiler that turn them into JavaScript function calls.

Sorry, but this full on ignores the main point. This introduces the first HTML non-compliant change. That means that all existing tooling that depends on HTML compliance of Angular templates needs to be changed/rewritten to support this feature.
You are willingly sacrifice the Angular HTML compliance of tempaltes, developers could count on till now, for a little bit of syntactic sugar?

I cite from the official Angular documentation itself: 
> Template syntax
> In Angular, a template is a chunk of HTML

e-oz commented:
@TillFProtzek current capabilities of tooling should not stop Angular evolution. Tools should adapt to changes and evolution, that's the reality. Somehow they adapted to standalone components, directive selectors, (event) [binding] syntax - they weren't stopped by the fact that it's not part of HTML.

kroeder commented:
What parser should people use when they write update schematics? We currently use parse5 for that. Is it possible to use the angular parser for that, does it have public api?

TillFProtzek commented:
@e-oz it's not the point if something was part or not part of HTML. All syntax extension of Angular are/were HTML syntax compliant. All points you mentioned are HTML syntax compliant. Templates until now could be parsed by any HTML-parser.

It seems that the full scope of that change is missed in regards of reward/risk. Formatters, Linters, sophisticated dev-tools - all have to change if the developers start to use the feature or adapt a template using it.

@samclaus the originator of the feature-request got convinced in the discussion there that it is perhaps not a good idea. So convinced that the only mention of the feature request afterwards, before this pull request out of the blue, is from himself. Pointing out again to what he learned about the risk of changing HTML compliance of the templates.

alfaproject commented:
I'm also not sure where this PR is coming from? Angular HTML template files are HTML-spec compliant and I believe that was the point of even using HTML to begin with as opposed to JSX or something else, right?

alxhub commented:
Hello all! I'm super excited for this feature and I love to see the discussion here.

Yes, this is a departure from templates being strictly HTML-compliant. We're fully aware of that - it's an intentional choice on our part, after thoughtful analysis of the tradeoffs involved. We're not making this decision purely for the syntactic sugar of `<my-cmp />`, either - there are good reasons to treat Angular templates as their own HTML-derived language.

## It enables some bigger DX improvements in templates

Syntactic sugar for self-closing tags is a small DX improvement of its own, but it's not the main motivation here. We're looking at a number of other improvements we can make to improve the template authoring experience. Things like improving the ergonomics of control flow, resolving longstanding inconsistencies in some of the binding syntax, and improving support for TypeScript as well as other tooling such as linters - all of which can benefit from native syntax in templates.

These are early thoughts, and will be bringing these kinds of changes to RFC as we nail down the details (although these things take time).

## Templates already are their own language

Today, Angular templates are syntactically HTML, but semantically distinct. Even if `[input]="expression"` is a valid syntax for an element attribute, it _means_ something different in Angular than it does in HTML. The same goes for `{{ 'interpolations' }}` - there's a whole sub-language inside of these syntaxes that's invisible to any tool which treats the template as plain HTML. It's true that simply parsing templates as HTML doesn't cause any problems, but it's easy for tooling to get into trouble when _interpreting_ the parsed template with HTML semantics - that can lead to subtle bugs.

Consider an HTML formatter running on the template expression `{{ data.replace('    ', '-') }}` (replacing a sequence of 4 spaces in some string with a dash). HTML formatters aren't aware of the switch in language context in what they consider to be plain text. In HTML, multiple spaces in a row can be collapsed into one, so the formatter may choose to "improve" this template by removing the extraneous spaces, unaware that it is altering the meaning of the template to Angular in the process. This may not seem like a huge problem, but it's an example of how treating Angular templates as plain HTML in tooling is fundamentally unsound today.

In practice, we've seen some tools avoid this problem by using Angular's own template parser to parse templates. The popular `angular-eslint` plugin does this, for example. In the future, we want to make it easier for other tools to do the same, and give the ecosystem APIs to work with Angular templates both syntactically and semantically.

## This specific change is minimally impactful

Self-closing tags are an optional language feature. If a project wants to enforce that its templates are parseable as plain HTML, it can do so (likely via a lint rule).

All tooling and parsers we've tested has been perfectly usable with self-closing component tags. In fact, they're often tripped up by other aspects of Angular syntax (case sensitivity, `[property]` and `(event)` bindings, etc).

## Summary

To be clear, we like HTML. Angular templates are HTML-flavored, and we have no plans to change that - writing templates should feel like writing HTML. But Angular templates _are_ special already, and the ergonomic benefits of relaxing strict compliance with the HTML spec from a syntactic standpoint are significant.

jelbourn reviewed (approved):
LGTM

@alxhub already mentioned most of what I would have said here.

The idea of "valid HTML templates" has historically been an important one to make sure that HTML documents rendered by the browser render as intended. Angular templates, though, are *never* rendered directly by the browser. An Angular template is always compiled into JavaScript instructions that manually create a DOM structure (via underlying APIs like `document.createElement`). While some developers may appreciate the concept of concept/experience of authoring spec-valid HTML, it does not actually have any bearing whatsoever on what's ultimately rendered into the DOM. 

This feature in particular is completely optional, and people can continue using closing tags for custom elements, so we don't see this as breaking. 

While we do expect to make changes to the template syntax over the years, we will always pay close attention to the stability and backwards of the ecosystem.

Yogu commented:
Thank you for these detailed responses! I see that you're not making this decision lightheartedly. I just would like to draw attention to the question asked by @kroeder again:

> What parser should people use when they write update schematics? We currently use parse5 for that. Is it possible to use the angular parser for that, does it have public api?

If library and tool authors are expected to support the new template syntax, it would be very helpful to provide a stable parser (and modification / printer) API. This is especially relevant when writing schematics that update templates which previously could use standard HTML parsers and printers. While a team developing an app can just decide not to use this new feature, library and tool authors are not in this luxurious position - if they want to claim say Angular 16 compatibility, they need to support the new template syntax.

crisbeto commented:
@alxhub I've reworked the PR so only non-native elements allow self-closing.

crisbeto committed:
feat(compiler): allow self-closing tags on custom elements

Allows for self-closing tags to be used for non-native tag names, e.g. `<foo [input]="bar"></foo>` can now be written as `<foo [input]="bar"/>`. Native tag names still have to have closing tags.

Fixes #39525.

alxhub commented on packages/compiler/src/selector.ts:
```diff
@@ -172,22 +170,6 @@ export class CssSelector {
     this.element = element;
   }
 
-  /** Gets a template string for an element that matches the selector. */
```
👍

bob-watson commented:
Is there any (new or update to existing) documentation to accompany this new feature?
I don't see any in this PR.

AndrewKushnir commented:
@crisbeto quick question: it looks like this change covers DOM elements, should we allow `<ng-content>` and `<ng-container>` to be self-closing as well (it's less applicable to `<ng-template>`)? I think it'd be great to allow this, especially for `<ng-content>` that doesn't have any child elements. This is **not** a blocker for this PR, we can create a followup PR if needed.

crisbeto commented:
@AndrewKushnir unless I'm missing something, this should allow `ng-content` and `ng-container` to be self-closing already. The change marks all known node types as NOT allowing self closing while all unknown ones (such as `ng-content`) can be self-closing.

AndrewKushnir commented:
@crisbeto yeah, you are right, thanks for the reply 👍

e-oz commented:
I hope "path" is not in the list of known tags - it would help SVG components.

alxhub commented:
This PR was merged into the repository by commit a532d71975bef463223fd5d8322e3140760c9134.

maximilian27 commented:
Nice, but how does this improve Angular in any way?

kroeder commented:
> Syntactic sugar for self-closing tags is a small DX improvement of its own, but it's not the main motivation here. We're looking at a number of other improvements we can make to improve the template authoring experience.

@maximilian27 See https://github.com/angular/angular/pull/48535#issuecomment-1357992123

> Syntactic sugar for self-closing tags is a small DX improvement of its own, but it's not the main motivation here. We're looking at a number of other improvements we can make to improve the template authoring experience.
